File: apps/cli/devx/internal/exposure/store.go
// Copyright (c) 2026 VitruvianSoftware
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

// Package exposure persists local metadata for active tunnel exposures
// that cannot be retrieved from the Cloudflare API (e.g. the target port).
package exposure

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

// Entry records metadata about a single exposed tunnel.
type Entry struct {
	TunnelName string `json:"tunnel_name"`
	TunnelID   string `json:"tunnel_id"`
	Port       string `json:"port"`
	Domain     string `json:"domain"`
}

// stateDir returns ~/.config/devx, creating it if necessary.
func stateDir() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	dir := filepath.Join(home, ".config", "devx")
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return "", err
	}
	return dir, nil
}

func statePath() (string, error) {
	dir, err := stateDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(dir, "exposures.json"), nil
}

// Save adds or updates an exposure entry.
func Save(entry Entry) error {
	entries, _ := LoadAll() // ignore read errors — start fresh

	// Upsert by tunnel name
	found := false
	for i, e := range entries {
		if e.TunnelName == entry.TunnelName {
			entries[i] = entry
			found = true
			break
		}
	}
	if !found {
		entries = append(entries, entry)
	}

	return writeAll(entries)
}

// Remove deletes all stored exposure entries.
func RemoveAll() error {
	return writeAll(nil)
}

// Remove deletes a single entry by tunnel name.
func RemoveByName(tunnelName string) error {
	entries, _ := LoadAll()
	var filtered []Entry
	for _, e := range entries {
		if e.TunnelName != tunnelName {
			filtered = append(filtered, e)
		}
	}
	return writeAll(filtered)
}

// LoadAll reads all persisted exposure entries.
func LoadAll() ([]Entry, error) {
	path, err := statePath()
	if err != nil {
		return nil, err
	}
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var entries []Entry
	if err := json.Unmarshal(data, &entries); err != nil {
		return nil, fmt.Errorf("parsing %s: %w", path, err)
	}
	return entries, nil
}

// LookupPort returns the port for a tunnel name, or "" if unknown.
func LookupPort(tunnelName string) string {
	entries, _ := LoadAll()
	for _, e := range entries {
		if e.TunnelName == tunnelName {
			return e.Port
		}
	}
	return ""
}

func writeAll(entries []Entry) error {
	path, err := statePath()
	if err != nil {
		return err
	}
	data, err := json.MarshalIndent(entries, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0o644)
}

File: apps/suites/tabula/web/lib/store.ts
/**
 * Copyright (c) 2026 VitruvianSoftware
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

"use client";

/**
 * Zustand store for web dashboard state management
 *
 * Provides centralized state for workspaces and stats,
 * consistent with the extension's state management patterns.
 */

import { create } from "zustand";
import type { Workspace, WorkspaceStats, DataSourceInfo } from "@/types";
import { AuthService, User } from "./auth";
import {
  getStats,
  getWorkspaces,
  getDataSourceInfo,
  seedDemoData,
} from "./data";

interface WorkspaceState {
  // Auth
  user: User | null;
  isAuthenticated: boolean;

  // Data
  workspaces: Workspace[];
  stats: WorkspaceStats | null;
  dataSource: DataSourceInfo | null;

  // Loading states
  loading: boolean;
  error: string | null;

  // Actions
  fetchWorkspaces: () => Promise<void>;
  fetchStats: () => Promise<void>;
  refresh: () => Promise<void>;
  seedDemoData: () => Promise<void>;
  login: () => Promise<void>;
  logout: () => void;
}

export const useWorkspaceStore = create<WorkspaceState>((set, get) => ({
  // Initial state
  user: AuthService.getUser(),
  isAuthenticated: !!AuthService.getToken(),
  workspaces: [],
  stats: null,
  dataSource: null,
  loading: true,
  error: null,

  // Fetch workspaces from data service
  fetchWorkspaces: async () => {
    try {
      set({ loading: true, error: null });
      const workspaces = await getWorkspaces();
      const dataSource = getDataSourceInfo();
      set({ workspaces, dataSource, loading: false });
    } catch (error) {
      set({
        error:
          error instanceof Error ? error.message : "Failed to load workspaces",
        loading: false,
      });
    }
  },

  // Fetch stats from data service
  fetchStats: async () => {
    try {
      set({ loading: true, error: null });
      set({ loading: true, error: null });
      const stats = await getStats();
      const dataSource = getDataSourceInfo();
      set({ stats, dataSource, loading: false });
      // set({ loading: false });
    } catch (error) {
      set({
        error: error instanceof Error ? error.message : "Failed to load stats",
        loading: false,
      });
    }
  },

  // Refresh all data
  refresh: async () => {
    const { fetchWorkspaces, fetchStats } = get();
    await Promise.all([fetchWorkspaces(), fetchStats()]);
  },

  // Seed demo data
  seedDemoData: async () => {
    seedDemoData();
    await get().refresh();
  },

  // Auth actions
  login: async () => {
    try {
      set({ loading: true, error: null });
      const { user } = await AuthService.login();
      set({ user, isAuthenticated: true, loading: false });

      // Refresh data after login
      const { refresh } = get();
      await refresh();
    } catch (error) {
      set({
        error: error instanceof Error ? error.message : "Login failed",
        loading: false,
      });
    }
  },

  logout: () => {
    AuthService.logout();
    set({
      user: null,
      isAuthenticated: false,
      workspaces: [],
      stats: null,
      dataSource: null,
    });
    // Refresh to update data source info (back to demo/none)
    get().fetchWorkspaces();
  },
}));
